### PR TITLE
Recreate cache of endpoints when data source updated

### PR DESCRIPTION
**Pull request by jbagga**
Addresses https://github.com/aspnet/Home/issues/2663

### CODE REVIEW

File: src/Microsoft.AspNetCore.Dispatcher/Tree/TreeMatcher.cs [hunk 1]
```diff
@@ -40,6 +37,7 @@ public TreeMatcher()
             }
 
             EnsureServicesInitialized(context);
+            Extensions.Primitives.ChangeToken.OnChange(() => ChangeToken, () => Volatile.Write(ref _dataInitialized, false));
```
Comment (jbagga) [inline src/Microsoft.AspNetCore.Dispatcher/Tree/TreeMatcher.cs]: I tried to do this in the ctor but the `DataSource` is set after the ctor in `TreeMatcherFactory` and `ChangeToken` was null as a result

File: src/Microsoft.AspNetCore.Dispatcher/Tree/TreeMatcher.cs [hunk 2]
```diff
@@ -136,8 +134,7 @@ internal Cache CreateCache()
             {
                 var endpoint = endpoints[i];
 
-                var templateEndpoint = endpoint as IRoutePatternEndpoint;
-                if (templateEndpoint == null)
+                if (!(endpoint is IRoutePatternEndpoint templateEndpoint))
```
Comment (jbagga) [inline src/Microsoft.AspNetCore.Dispatcher/Tree/TreeMatcher.cs]: We can thank @pranavkm for this

File: test/Microsoft.AspNetCore.Dispatcher.Test/Tree/TreeMatcherTest.cs [hunk 3]
```diff
@@ -776,6 +780,32 @@ public void MatchAsync_DoesNotMatchShorterUrl()
...
+
+            // Arrange 2
+            dataSource.UpdateEndpoints(new List<Endpoint>() { new RoutePatternEndpoint("test2/{p1}", new object(), Test_Delegate, "Test2"), });
+            context = CreateMatcherContext("/test2/parameter2");
+
+            // Act 2
+            await matcher.MatchAsync(context);
+
+            // Assert 2
+            Assert.Equal("Test2", context.Endpoint.DisplayName);
```
Comment (jbagga) [inline test/Microsoft.AspNetCore.Dispatcher.Test/Tree/TreeMatcherTest.cs]: I wanted to check if context.Values are populated with the key "p1" with value "parameter2" but context.Values is an empty collection. Is that expected?

File: src/Microsoft.AspNetCore.Dispatcher/CompositeDispatcherDataSource.cs [hunk 4]
```diff
@@ -32,7 +32,7 @@ public CompositeDispatcherDataSource(IEnumerable<DispatcherDataSource> dataSourc
 
         public override IChangeToken ChangeToken { get; }
 
-        protected override IReadOnlyList<Address> GetAddesses()
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Dispatcher/CompositeDispatcherDataSource.cs]: lol

File: src/Microsoft.AspNetCore.Dispatcher/Tree/TreeMatcher.cs (hunk 1 above)
Comment (rynowak) [reply src/Microsoft.AspNetCore.Dispatcher/Tree/TreeMatcher.cs]: You could do this in `EnsureServicesInitialized`. You can't really do it here, it's going to add another delegate every request.

Comment (rynowak) [reply src/Microsoft.AspNetCore.Dispatcher/Tree/TreeMatcher.cs]: Can you add a using for this?

File: src/Microsoft.AspNetCore.Dispatcher/Tree/TreeMatcher.cs (hunk 2 above)
Comment (rynowak) [reply src/Microsoft.AspNetCore.Dispatcher/Tree/TreeMatcher.cs]: cool beans

File: test/Microsoft.AspNetCore.Dispatcher.Test/Tree/TreeMatcherTest.cs (hunk 3 above)
Comment (rynowak) [reply test/Microsoft.AspNetCore.Dispatcher.Test/Tree/TreeMatcherTest.cs]: No, that doesn't seem right.

File: test/Microsoft.AspNetCore.Dispatcher.Test/Tree/TreeMatcherTest.cs [hunk 5]
```diff
@@ -804,5 +834,65 @@ private static Task Test_Delegate(HttpContext httpContext)
...
+                _changeToken = new TestChangeToken();
+            }
+
+            public void UpdateEndpoints(IList<Endpoint> endpoints)
+            {
+                lock (_lock)
+                {
+                    _endpoints = endpoints;
+                    // Trigger change token here
+                    var previousToken = Interlocked.Exchange(ref _changeToken, new TestChangeToken());
```
Comment (rynowak) [inline test/Microsoft.AspNetCore.Dispatcher.Test/Tree/TreeMatcherTest.cs]: The interlocked isn't really needed here because it's protected by a lock.

**Comment by jbagga** [comment]
@rynowak  Updated

**Pull request merged**